driver/driver.go

```diff
@@ -787,6 +787,12 @@ type unwrappable interface {
 	Unwrap() error
 }
 
+// TODO driver.ErrBadConn should not be returned when there's a possibility that
+// the query has been executed. In our case there is a window in protocol.Call
```

Comment:
It probably goes without saying, but just to be sure we are on the same page: this issue should be dealt with by implementing session support in the dqlite state machine, as per paragraph 6.3 of the raft dissertation.

Until then, I'm fine converting `io.EOF` into `driver.ErrBadConn` if it turns out that it solves more problems than it creates (i.e. the chance of returning `driver.ErrBadConn` also if a query was executed is low, while the chance that converting `io.EOF` into `driver.ErrBadConn` solves actual real world issues by retrying is high).

Still, I would have been happier if there was an end-to-end test for this change, showing that we at least precisely understand and can reproduce the exact sequence of events around the issue. With such a test in place it would be easier to avoid regressions later down the road, especially when we'll implement sessions as per 6.3.